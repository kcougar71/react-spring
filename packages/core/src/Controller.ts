import {
  is,
  each,
  OneOrMore,
  toArray,
  UnknownProps,
  FluidProps,
  Remap,
} from 'shared'
import * as G from 'shared/globals'

import { SpringValues, RangeProps } from './types/spring'
import {
  OnProps,
  OnRest,
  OnStart,
  OnChange,
  AnimationResult,
  AnimationProps,
  LoopProp,
} from './types/animated'
import { Indexable, Falsy } from './types/common'
import { runAsync, scheduleProps, RunAsyncState, AsyncResult } from './runAsync'
import { SpringPhase, CREATED, ACTIVE, IDLE } from './SpringPhase'
import { inferTo } from './helpers'
import { SpringValue, getPropsForLoop } from './SpringValue'
import { FrameValue } from './FrameValue'

/** Events batched by the `Controller` class */
const BATCHED_EVENTS = ['onStart', 'onChange', 'onRest'] as const

/** All event props supported by the `Controller` class */
export type EventProps<State extends object> = {
  /**
   * Called when the # of animating values exceeds 0
   *
   * Also accepts an object for per-key events
   */
  onStart?: (() => void) | { [P in keyof State]?: OnStart<State[P]> }
  /**
   * Called when the # of animating values hits 0
   *
   * Also accepts an object for per-key events
   */
  onRest?: OnRest<State> | { [P in keyof State]?: OnRest<State[P]> }
  /**
   * Called after an animation is updated by new props.
   * Useful for manipulation
   *
   * Also accepts an object for per-key events
   */
  onProps?: OnProps<State> | { [P in keyof State]?: OnProps<State[P]> }
  /**
   * Called once per frame when animations are active
   *
   * Also accepts an object for per-key events
   */
  onChange?:
    | ((values: State) => void)
    | { [P in keyof State]?: OnChange<State[P]> }
}

export type ControllerProps<State extends Indexable = UnknownProps> = Remap<
  FluidProps<Partial<State>> &
    RangeProps<State> &
    EventProps<State> &
    AnimationProps & {
      loop?: LoopProp<ControllerProps<State>>
    }
>

/** An update that hasn't been applied yet */
type PendingProps<State extends Indexable> = ControllerProps<State> & {
  /** The keys affected by this update. When null, all keys are affected. */
  keys: string[] | null
}

/** The flush function that handles `start` calls */
export type FlushFn<State extends Indexable> = (
  ctrl: Controller<State>,
  queue: PendingProps<State>[]
) => AsyncResult<State>

let nextId = 1
let lastAsyncId = 0

export class Controller<State extends Indexable = UnknownProps>
  implements FrameValue.Observer {
  readonly id = nextId++

  /** The animated values */
  springs: SpringValues<State> = {} as any

  /** The queue of pending props */
  queue: PendingProps<State>[] = []

  /** Custom handler for flushing update queues */
  protected _flush?: FlushFn<State>

  /** These props are used by all future spring values */
  protected _initialProps?: Indexable

  /** The combined phase of our spring values */
  protected _phase: SpringPhase = CREATED

  /** The values currently being animated */
  protected _active = new Set<FrameValue>()

  /** State used by the `runAsync` function */
  protected _state: RunAsyncState<State> = {}

  /** The event queues that are flushed once per frame maximum */
  protected _events = {
    onStart: new Set<Function>(),
    onChange: new Set<Function>(),
    onRest: new Map<OnRest, AnimationResult>(),
  }

  constructor(props?: ControllerProps<State> | null, flush?: FlushFn<State>) {
    this._onFrame = this._onFrame.bind(this)
    if (flush) {
      this._flush = flush
    }
    if (props) {
      const { to, ...initialProps } = inferTo(props)
      this._initialProps = initialProps
      if (to) this.start({ to } as any)
    }
  }

  /**
   * Equals `true` when no spring values are in the frameloop, and
   * no async animation is currently active.
   */
  get idle() {
    return (
      !this._state.promise &&
      Object.values(this.springs as Indexable<SpringValue>).every(
        spring => spring.idle
      )
    )
  }

  /** Get the current values of our springs */
  get(): State & UnknownProps {
    const values: any = {}
    this.each((spring, key) => (values[key] = spring.get()))
    return values
  }

  /** Push an update onto the queue of each value. */
  update(props: ControllerProps<State> | Falsy) {
    if (props) this.queue.push(createUpdateProps(props))
    return this
  }

  /**
   * Start the queued animations for every spring, and resolve the returned
   * promise once all queued animations have finished or been cancelled.
   *
   * When you pass a queue (instead of nothing), that queue is used instead of
   * the queued animations added with the `update` method, which are left alone.
   */
  start(props?: OneOrMore<ControllerProps<State>> | null): AsyncResult<State> {
    if (!props) {
      props = this.queue
      this.queue = []
    }
    const queue = toArray<any>(props).map(createUpdateProps)
    if (this._flush) {
      return this._flush(this, queue)
    }
    each(queue, props => {
      prepareSprings(this.springs, props, key => {
        return createSpring(key, this._initialProps, this)
      })
    })
    return flushUpdateQueue(this, queue)
  }

  /** Stop one animation, some animations, or all animations */
  stop(keys?: OneOrMore<string>) {
    if (is.und(keys)) {
      this.each(spring => spring.stop())
    } else {
      const springs = this.springs as Indexable<SpringValue>
      each(toArray(keys), key => springs[key].stop())
    }
    return this
  }

  /** Freeze the active animation in time */
  pause(keys?: OneOrMore<string>) {
    if (is.und(keys)) {
      this.each(spring => spring.pause())
    } else {
      const springs = this.springs as Indexable<SpringValue>
      each(toArray(keys), key => springs[key].pause())
    }
    return this
  }

  /** Resume the animation if paused. */
  resume(keys?: OneOrMore<string>) {
    if (is.und(keys)) {
      this.each(spring => spring.resume())
    } else {
      const springs = this.springs as Indexable<SpringValue>
      each(toArray(keys), key => springs[key].resume())
    }
    return this
  }

  /** Restart every animation. */
  reset() {
    this.each(spring => spring.reset())
    // TODO: restart async "to" prop
    return this
  }

  /** Call a function once per spring value */
  each(iterator: (spring: SpringValue, key: string) => void) {
    each(this.springs, iterator as any)
  }

  /** Destroy every spring in this controller */
  dispose() {
    this._state.asyncTo = undefined
    this.each(spring => spring.dispose())
    this.springs = {} as any
  }

  /** @internal Called at the end of every animation frame */
  protected _onFrame() {
    const { onStart, onChange, onRest } = this._events

    const isActive = this._active.size > 0
    if (isActive && this._phase != ACTIVE) {
      this._phase = ACTIVE
      flush(onStart, onStart => onStart())
    }

    const values = (onChange.size || (!isActive && onRest.size)) && this.get()
    flush(onChange, onChange => onChange(values))

    // The "onRest" queue is only flushed when all springs are idle.
    if (!isActive) {
      this._phase = IDLE
      flush(onRest, (result, onRest) => {
        result.value = values
        onRest(result)
      })
    }
  }

  /** @internal */
  onParentChange(event: FrameValue.Event) {
    if (event.type == 'change') {
      this._active[event.idle ? 'delete' : 'add'](event.parent)
      G.frameLoop.onFrame(this._onFrame)
    }
  }
}

/** Find keys with defined values */
function findDefined(values: any, keys: Set<string>) {
  each(values, (value, key) => value != null && keys.add(key as any))
}

/** Basic helper for clearing a queue after processing it */
function flush<P, T>(
  queue: Map<P, T>,
  iterator: (value: T, key: P) => void
): void
function flush<T>(queue: Set<T>, iterator: (value: T) => void): void
function flush(queue: any, iterator: any) {
  if (queue.size) {
    each(queue, iterator)
    queue.clear()
  }
}

/**
 * Warning: Props might be mutated.
 */
export function flushUpdateQueue(
  ctrl: Controller<any>,
  queue: PendingProps<any>[]
) {
  return Promise.all(queue.map(props => flushUpdate(ctrl, props))).then(
    results => ({
      value: ctrl.get(),
      finished: results.every(finished => finished),
    })
  )
}

/**
 * Warning: Props might be mutated.
 *
 * Process a single set of props using the given controller.
 *
 * The returned promise resolves to `true` once the update is
 * applied and any animations it starts are finished without being
 * stopped or cancelled.
 */
export function flushUpdate(
  ctrl: Controller<any>,
  props: PendingProps<any>
): Promise<boolean> {
  const { to, loop } = props

  const asyncTo = is.arr(to) || is.fun(to) ? to : undefined
  if (asyncTo) {
    props.to = undefined
  }

  // Looping must be handled in this function, or else the values
  // would end up looping out-of-sync in many common cases.
  if (loop) {
    props.loop = false
  }

  // Batched events are queued by individual spring values.
  each(BATCHED_EVENTS, key => {
    const handler: any = props[key]
    if (is.fun(handler)) {
      const queue = ctrl['_events'][key]
      props[key] =
        queue instanceof Set
          ? () => queue.add(handler)
          : ((({ finished, cancelled }: AnimationResult) => {
              const result = queue.get(handler)
              if (result) {
                if (!finished) result.finished = false
                if (cancelled) result.cancelled = true
              } else {
                // The "value" is set before the "handler" is called.
                queue.set(handler, {
                  value: null,
                  finished,
                  cancelled,
                })
              }
            }) as any)
    }
  })

  const keys = props.keys || Object.keys(ctrl.springs)
  const promises = keys.map(key => ctrl.springs[key]!.start(props as any))

  // Schedule the "asyncTo" if defined.
  if (asyncTo) {
    const state = ctrl['_state']
    promises.push(
      scheduleProps(++lastAsyncId, {
        props,
        state,
        action(props, resolve) {
          // Start, replace, or cancel the async animation.
          props.onRest = undefined
          resolve(
            runAsync<any>(
              asyncTo,
              props,
              state,
              ctrl.get.bind(ctrl),
              () => false, // TODO: add pausing to Controller
              ctrl.start.bind(ctrl) as any,
              ctrl.stop.bind(ctrl)
            )
          )
        },
      })
    )
  }

  return Promise.all(promises).then(results => {
    const finished = results.every(result => result.finished)
    if (finished && loop) {
      const nextProps = getPropsForLoop(props, loop, to)
      if (nextProps) {
        return flushUpdate(ctrl, nextProps)
      }
    }
    return finished
  })
}

export function createUpdateProps(props: any) {
  if ('keys' in props) return props
  const { to, from } = (props = inferTo(props))

  // Collect the keys affected by this update.
  const keys = new Set<string>()

  if (from) {
    findDefined(from, keys)
  } else {
    // Falsy values are ignored.
    props.from = undefined
  }

  if (is.obj(to)) {
    findDefined(to, keys)
  } else if (!to) {
    // Falsy values are ignored.
    props.to = undefined
  }

  props.keys = keys.size ? Array.from(keys) : null
  return props
}

/**
 * From an array of updates, get the map of `SpringValue` objects
 * by their keys. Springs are created when any update wants to
 * animate a new key.
 */
export function getSprings<State>(
  ctrl: Controller<State>,
  props?: OneOrMore<ControllerProps<State>>
) {
  const springs = { ...ctrl.springs }
  if (props) {
    const initialProps = ctrl['_initialProps']
    each(toArray(props), (props: any) => {
      props = createUpdateProps(props)
      if (!is.obj(props.to)) {
        // Avoid passing array/function to each spring.
        props = { ...props, to: undefined }
      }
      prepareSprings(springs, props, key => {
        return createSpring(key, initialProps)
      })
    })
  }
  return springs
}

/**
 * Tell a controller to manage the given `SpringValue` objects
 * whose key is not already in use.
 */
export function setSprings(
  ctrl: Controller,
  springs: SpringValues<UnknownProps>
) {
  each(springs, (spring, key) => {
    if (!ctrl.springs[key]) {
      ctrl.springs[key] = spring
      spring.addChild(ctrl)
    }
  })
}

function createSpring(
  key: string,
  initialProps?: UnknownProps,
  observer?: FrameValue.Observer
) {
  const spring = new SpringValue(initialProps)
  spring.key = key
  if (observer) {
    spring.addChild(observer)
  }
  return spring
}

function prepareSprings(
  springs: any,
  props: PendingProps<any>,
  create: (key: string) => SpringValue
) {
  if (props.keys) {
    each(props.keys, key => {
      const spring = springs[key] || (springs[key] = create(key))
      spring['_prepareNode'](props)
    })
  }
}
