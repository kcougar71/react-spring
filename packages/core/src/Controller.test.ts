import { Controller } from './Controller'

describe('Controller', () => {
  it('can animate a number', () => {
    const ctrl = new Controller({ x: 0 })
    ctrl.start({ x: 100 })

    const frames = getFrames(ctrl)
    expect(frames).toMatchSnapshot()

    // The first frame should *not* be the from value.
    expect(frames[0]).not.toEqual({ x: 0 })

    // The last frame should be the goal value.
    expect(frames.slice(-1)[0]).toEqual({ x: 100 })
  })

  it('can animate an array of numbers', () => {
    const config = { precision: 0.005 }
    const ctrl = new Controller<{ x: [number, number] }>({ x: [1, 2], config })
    ctrl.start({ x: [5, 10] })

    const frames = getFrames(ctrl)
    expect(frames).toMatchSnapshot()

    // The last frame should be the goal value.
    expect(frames.slice(-1)[0]).toEqual({ x: [5, 10] })

    // The 2nd value is always ~2x the 1st value (within the defined precision).
    const factors = frames.map(frame => frame.x[1] / frame.x[0])
    expect(
      factors.every(factor => Math.abs(2 - factor) < config.precision)
    ).toBeTruthy()
  })

  describe('when the "to" prop is an async function', () => {
    it('acts strangely without the "from" prop', async () => {
      const ctrl = new Controller<{ x: number }>()

      const { springs } = ctrl
      ctrl.start({
        to: async update => {
          // The spring does not exist yet!
          expect(springs.x).toBeUndefined()

          // Any values passed here are treated as "from" values,
          // because no "from" prop was ever given.
          update({ x: 1 })
          // Now the spring exists!
          expect(springs.x).toBeDefined()
          // But the spring is idle!
          expect(springs.x.idle).toBeTruthy()

          // This call *will* start an animation!
          update({ x: 2 })
          expect(springs.x.idle).toBeFalsy()
        },
      })

      // Since we call `update` twice, frames are generated!
      expect(await getAsyncFrames(ctrl)).toMatchSnapshot()
    })
  })

  describe('when the "onStart" prop is defined', () => {
    it('is called once per "start" call maximum', () => {
      const ctrl = new Controller({ x: 0, y: 0 })

      const onStart = jest.fn()
      ctrl.start({
        x: 1,
        y: 1,
        onStart,
      })

      advanceUntilIdle()
      expect(onStart).toBeCalledTimes(1)
    })

    it('can be different per key', () => {
      const ctrl = new Controller({ x: 0, y: 0 })

      const onStart1 = jest.fn()
      ctrl.start({ x: 1, onStart: onStart1 })

      const onStart2 = jest.fn()
      ctrl.start({ y: 1, onStart: onStart2 })

      advanceUntilIdle()
      expect(onStart1).toBeCalledTimes(1)
      expect(onStart2).toBeCalledTimes(1)
    })
  })
})
