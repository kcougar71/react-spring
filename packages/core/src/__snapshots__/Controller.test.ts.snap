// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controller can animate a number 1`] = `
Array [
  Object {
    "x": 2.263484330785799,
  },
  Object {
    "x": 7.615528894993947,
  },
  Object {
    "x": 14.72729545687759,
  },
  Object {
    "x": 22.67933293781441,
  },
  Object {
    "x": 30.85053191603684,
  },
  Object {
    "x": 38.83524421610096,
  },
  Object {
    "x": 46.38171205392548,
  },
  Object {
    "x": 53.346574480369156,
  },
  Object {
    "x": 59.66146523130329,
  },
  Object {
    "x": 65.30867180317823,
  },
  Object {
    "x": 70.30355736424129,
  },
  Object {
    "x": 74.68200657416291,
  },
  Object {
    "x": 78.49158337229619,
  },
  Object {
    "x": 81.78541406742545,
  },
  Object {
    "x": 84.61805634019797,
  },
  Object {
    "x": 87.04280232770594,
  },
  Object {
    "x": 89.11000586164683,
  },
  Object {
    "x": 90.8661309883985,
  },
  Object {
    "x": 92.35329941206051,
  },
  Object {
    "x": 93.60917483504707,
  },
  Object {
    "x": 94.66706719880813,
  },
  Object {
    "x": 95.55617327559861,
  },
  Object {
    "x": 96.30189477450246,
  },
  Object {
    "x": 96.92619326731115,
  },
  Object {
    "x": 97.4479544590839,
  },
  Object {
    "x": 97.88334387327176,
  },
  Object {
    "x": 98.2461428371077,
  },
  Object {
    "x": 98.54805845242963,
  },
  Object {
    "x": 98.7990045563195,
  },
  Object {
    "x": 99.0073529166361,
  },
  Object {
    "x": 99.18015536949562,
  },
  Object {
    "x": 99.3233385117055,
  },
  Object {
    "x": 99.4418730756026,
  },
  Object {
    "x": 99.53992035746593,
  },
  Object {
    "x": 99.62095813165256,
  },
  Object {
    "x": 99.68788842438258,
  },
  Object {
    "x": 99.74312938902625,
  },
  Object {
    "x": 99.78869335076726,
  },
  Object {
    "x": 99.8262528947063,
  },
  Object {
    "x": 99.85719667273446,
  },
  Object {
    "x": 99.8826764105673,
  },
  Object {
    "x": 100,
  },
]
`;

exports[`Controller can animate an array of numbers 1`] = `
Array [
  Object {
    "x": Array [
      1.0905393732314317,
      2.1810787464628634,
    ],
  },
  Object {
    "x": Array [
      1.3046211557997573,
      2.6092423115995147,
    ],
  },
  Object {
    "x": Array [
      1.589091818275103,
      3.178183636550206,
    ],
  },
  Object {
    "x": Array [
      1.9071733175125758,
      3.8143466350251516,
    ],
  },
  Object {
    "x": Array [
      2.2340212766414735,
      4.468042553282947,
    ],
  },
  Object {
    "x": Array [
      2.5534097686440376,
      5.106819537288075,
    ],
  },
  Object {
    "x": Array [
      2.8552684821570185,
      5.710536964314037,
    ],
  },
  Object {
    "x": Array [
      3.1338629792147654,
      6.267725958429531,
    ],
  },
  Object {
    "x": Array [
      3.3864586092521307,
      6.772917218504261,
    ],
  },
  Object {
    "x": Array [
      3.612346872127128,
      7.224693744254256,
    ],
  },
  Object {
    "x": Array [
      3.8121422945696497,
      7.624284589139299,
    ],
  },
  Object {
    "x": Array [
      3.987280262966515,
      7.97456052593303,
    ],
  },
  Object {
    "x": Array [
      4.139663334891849,
      8.279326669783698,
    ],
  },
  Object {
    "x": Array [
      4.271416562697019,
      8.542833125394038,
    ],
  },
  Object {
    "x": Array [
      4.3847222536079205,
      8.769444507215841,
    ],
  },
  Object {
    "x": Array [
      4.481712093108238,
      8.963424186216477,
    ],
  },
  Object {
    "x": Array [
      4.5644002344658725,
      9.128800468931745,
    ],
  },
  Object {
    "x": Array [
      4.634645239535937,
      9.269290479071874,
    ],
  },
  Object {
    "x": Array [
      4.694131976482419,
      9.388263952964838,
    ],
  },
  Object {
    "x": Array [
      4.7443669934018855,
      9.488733986803771,
    ],
  },
  Object {
    "x": Array [
      4.786682687952328,
      9.573365375904656,
    ],
  },
  Object {
    "x": Array [
      4.822246931023949,
      9.644493862047899,
    ],
  },
  Object {
    "x": Array [
      4.852075790980104,
      9.704151581960208,
    ],
  },
  Object {
    "x": Array [
      4.877047730692452,
      9.754095461384903,
    ],
  },
  Object {
    "x": Array [
      4.89791817836336,
      9.79583635672672,
    ],
  },
  Object {
    "x": Array [
      4.915333754930874,
      9.830667509861748,
    ],
  },
  Object {
    "x": Array [
      4.929845713484313,
      9.859691426968626,
    ],
  },
  Object {
    "x": Array [
      4.941922338097189,
      9.883844676194379,
    ],
  },
  Object {
    "x": Array [
      4.951960182252786,
      9.903920364505572,
    ],
  },
  Object {
    "x": Array [
      4.96029411666545,
      9.9205882333309,
    ],
  },
  Object {
    "x": Array [
      4.967206214779832,
      9.934412429559664,
    ],
  },
  Object {
    "x": Array [
      4.972933540468227,
      9.945867080936454,
    ],
  },
  Object {
    "x": Array [
      4.97767492302411,
      9.95534984604822,
    ],
  },
  Object {
    "x": Array [
      4.981596814298641,
      9.963193628597281,
    ],
  },
  Object {
    "x": Array [
      4.984838325266105,
      9.96967665053221,
    ],
  },
  Object {
    "x": Array [
      4.9875155369753035,
      9.975031073950607,
    ],
  },
  Object {
    "x": Array [
      4.989725175561048,
      9.979450351122097,
    ],
  },
  Object {
    "x": Array [
      4.991547734030691,
      9.983095468061382,
    ],
  },
  Object {
    "x": Array [
      4.993050115788252,
      9.986100231576504,
    ],
  },
  Object {
    "x": Array [
      4.994287866909378,
      9.988575733818756,
    ],
  },
  Object {
    "x": Array [
      4.995027593476066,
      9.990614112845385,
    ],
  },
  Object {
    "x": Array [
      4.995027593476066,
      9.992291713056268,
    ],
  },
  Object {
    "x": Array [
      4.995027593476066,
      9.99367173585381,
    ],
  },
  Object {
    "x": Array [
      4.995027593476066,
      9.994806458630702,
    ],
  },
  Object {
    "x": Array [
      5,
      10,
    ],
  },
]
`;

exports[`Controller the "loop" prop can be combined with the "reverse" prop 1`] = `
Array [
  0.33334,
  0.66668,
  1,
  0.66666,
  0.33331999999999995,
  0,
  0.33334,
  0.66668,
  1,
]
`;

exports[`Controller when the "to" prop is an async function acts strangely without the "from" prop 1`] = `
Array [
  Object {
    "x": 1.0226348433078583,
  },
  Object {
    "x": 1.0761552889499395,
  },
  Object {
    "x": 1.1472729545687759,
  },
  Object {
    "x": 1.226793329378144,
  },
  Object {
    "x": 1.3085053191603682,
  },
  Object {
    "x": 1.3883524421610094,
  },
  Object {
    "x": 1.4638171205392543,
  },
  Object {
    "x": 1.5334657448036912,
  },
  Object {
    "x": 1.5966146523130327,
  },
  Object {
    "x": 1.6530867180317823,
  },
  Object {
    "x": 1.7030355736424132,
  },
  Object {
    "x": 1.74682006574163,
  },
  Object {
    "x": 1.7849158337229636,
  },
  Object {
    "x": 1.817854140674256,
  },
  Object {
    "x": 1.8461805634019814,
  },
  Object {
    "x": 1.870428023277061,
  },
  Object {
    "x": 1.8911000586164695,
  },
  Object {
    "x": 1.9086613098839855,
  },
  Object {
    "x": 1.9235329941206056,
  },
  Object {
    "x": 1.9360917483504718,
  },
  Object {
    "x": 1.9466706719880824,
  },
  Object {
    "x": 1.9555617327559878,
  },
  Object {
    "x": 1.9630189477450264,
  },
  Object {
    "x": 1.9692619326731133,
  },
  Object {
    "x": 1.9744795445908405,
  },
  Object {
    "x": 1.9788334387327189,
  },
  Object {
    "x": 1.9824614283710786,
  },
  Object {
    "x": 1.9854805845242978,
  },
  Object {
    "x": 1.9879900455631967,
  },
  Object {
    "x": 1.9900735291663625,
  },
  Object {
    "x": 1.991801553694958,
  },
  Object {
    "x": 1.9932333851170567,
  },
  Object {
    "x": 1.9944187307560275,
  },
  Object {
    "x": 1.9953992035746602,
  },
  Object {
    "x": 1.9962095813165261,
  },
  Object {
    "x": 1.9968788842438259,
  },
  Object {
    "x": 1.997431293890262,
  },
  Object {
    "x": 1.9978869335076728,
  },
  Object {
    "x": 1.998262528947063,
  },
  Object {
    "x": 1.9985719667273445,
  },
  Object {
    "x": 1.9988267641056732,
  },
  Object {
    "x": 2,
  },
]
`;
